### PR TITLE
Add support for configurable set of immutable commits

### PR DESCRIPTION
**Pull request by martinvonz**
I moved most of #2247 into this, since this bit seems less controversial, and I'm not sure when I'll get back to working on the remaining parts (adding `immutable()` and using that in the default log template)

# Checklist

If applicable:
- [x] I have updated `CHANGELOG.md`
- [x] I have updated the documentation (README.md, docs/, demos/)
- [x] I have updated the config schema (cli/src/config-schema.json)
- [x] I have added tests to cover my changes

### CODE REVIEW

File: cli/src/cli_util.rs [hunk 1]
```diff
@@ -1259,16 +1259,29 @@ Set which revision the branch points to with `jj branch set {branch_name} -r <RE
...
-        let invalid_revset = to_rewrite_revset.intersection(&immutable_revset);
-        let revset = self.evaluate_revset(invalid_revset)?;
+        let (params, immutable_heads_str) = self
+            .revset_aliases_map
+            .get_function("immutable_heads")
+            .unwrap();
+        if !params.is_empty() {
+            return Err(user_error(
+                r#"The `revset-aliases.immutable_heads()` function must be declared
+without arguments."#,
```
Comment (yuja) [inline cli/src/cli_util.rs:1269]: nit: useless newline in error message?

**Review by yuja — APPROVED** [review]
Thanks!

> I'll get back to working on the remaining parts (adding immutable() and using that in the default log template)

I think the log revset can be defined without using `immutable()`.

File: cli/src/cli_util.rs (hunk 1 above)
Comment (martinvonz) [reply cli/src/cli_util.rs]: Done.

**Comment by martinvonz** [comment]
> I think the log revset can be defined without using `immutable()`.

True, and there are some advantages to doing it that way too, as you said (makes it easier for the revset engine to optimize). Maybe I'll put that in its own little PR.

**Review by rslabbert — APPROVED** [review]
Looks good, seeing this in action I'm really loving it and I think it has some great applicability!

**Pull request merged**